File: python/twelve-days/twelve_days.py
LYRICS = {
    'first': 'a Partridge in a Pear Tree.',
    'second': 'two Turtle Doves,',
    'third': 'three French Hens,',
    'fourth': 'four Calling Birds,',
    'fifth': 'five Gold Rings,',
    'sixth': 'six Geese-a-Laying,',
    'seventh': 'seven Swans-a-Swimming,',
    'eighth': 'eight Maids-a-Milking,',
    'ninth': 'nine Ladies Dancing,',
    'tenth': 'ten Lords-a-Leaping,',
    'eleventh': 'eleven Pipers Piping,',
    'twelfth': 'twelve Drummers Drumming,'
}


def recite(start_verse, end_verse):
    all_lyrics = []
    for n in range(start_verse, end_verse + 1):
        if n == 1:
            lyrics = 'On the first day of Christmas my true love gave to me: {}'.format(LYRICS['first'])
        else:
            lyrics_list = [['first', 'and ' + LYRICS['first']]]
            for i, e in enumerate(LYRICS.items()):
                if i >= n:
                    break
                elif i > 0:
                    lyrics_list.append(e)

            lyrics = ''
            last_lyrics = lyrics_list.pop()
            for i in range(0, len(lyrics_list)):
                ly = lyrics_list[-1-i]
                lyrics = lyrics + ' ' + ly[-1]
            lyrics = 'On the {} day of Christmas my true love gave to me: {}'.format(last_lyrics[0], last_lyrics[-1]) + lyrics

        all_lyrics.append(lyrics)
    return all_lyrics

File: ruby/twelve-days/twelve_days.rb
class TwelveDays

  def self.song
    weekdays = %w(first second third fourth fifth sixth seventh eighth ninth tenth eleventh twelfth)
    lyrics = []

    weekdays.each_with_index do |w, i|
      lyrics << lyric_for_line(w, i)
    end
    lyrics.join("\n")
  end

  def self.lyric_for_line(weekday, line_number)
    items = ["twelve Drummers Drumming", "eleven Pipers Piping", "ten Lords-a-Leaping", "nine Ladies Dancing", "eight Maids-a-Milking", "seven Swans-a-Swimming", "six Geese-a-Laying", "five Gold Rings", "four Calling Birds", "three French Hens", "two Turtle Doves", "a Partridge in a Pear Tree"].reverse

    lyric = ["On the #{weekday} day of Christmas my true love gave to me, "]
    current_line_items = items.select.with_index { |e, i| i <= line_number }
    current_line_items.reverse.each_with_index do |item, i|
      lyric << ( i + 1 == current_line_items.size ? (current_line_items.size > 1 ? "and " : '') + item + ".\n"
                                                  : item + ", " )
    end
    lyric.join
  end

end

module BookKeeping
  VERSION = 2
end